From 1ada0f4c2bf50904a70e484181aae26c8eafba01 Mon Sep 17 00:00:00 2001
From: Hans Christian Lonstad <hcl@datarespons.no>
Date: Thu, 23 Aug 2018 13:08:43 +0200
Subject: [PATCH 09/11] imx serial driver: Add DT option for non DMA

Based on stability issues observed using DMA, add a DT option to turn
off DMA
---
 drivers/tty/serial/imx.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/tty/serial/imx.c b/drivers/tty/serial/imx.c
index 4e827e5a52a3..0a6ae1aa8def 100644
--- a/drivers/tty/serial/imx.c
+++ b/drivers/tty/serial/imx.c
@@ -228,6 +228,7 @@ struct imx_port {
 	unsigned int		dma_tx_nents;
 	unsigned int            saved_reg[10];
 	bool			context_saved;
+	bool			nodma;
 };
 
 struct imx_port_ucrs {
@@ -1269,7 +1270,10 @@ static int imx_startup(struct uart_port *port)
 	writel(temp & ~UCR4_DREN, sport->port.membase + UCR4);
 
 	/* Can we enable the DMA support? */
-	if (!uart_console(port) && !sport->dma_is_inited)
+
+	if (sport->nodma)
+		dev_info(port->dev, "DMA disabled in DT\n");
+	if (!sport->nodma && !uart_console(port) && !sport->dma_is_inited)
 		imx_uart_dma_init(sport);
 
 	spin_lock_irqsave(&sport->port.lock, flags);
@@ -2053,6 +2057,9 @@ static int serial_imx_probe_dt(struct imx_port *sport,
 	if (of_get_property(np, "rts-gpios", NULL))
 		sport->have_rtsgpio = 1;
 
+	if (of_get_property(np, "nodma", NULL))
+		sport->nodma = 1;
+
 	return 0;
 }
 #else
-- 
2.17.1

